CODING_STANDARDS.md

```diff
@@ -0,0 +1,137 @@
+# Coding Standards
+
+This is a working document outlining the coding standard for the general MariaDB codebase. It does not cover the coding standards for individual plugins, these should have their own coding standards documentation.
+
+## Coding Style (C / C++ files)
+
+Everyone has a preferred coding style, there is no real correct style for all projects around the world. What is important is that we stick to one common style throughout this code base.
+
+### Indentation
+
+We should use a variant of the Allman indentation style. The variation is to use 2 spaces instead of tabs.
+
+Allman style specifies that braces associated with a statement should be on the following line with the same indentation and the statements inside the braces are next level indented. The closing braces are also on a new line at the same indentation as the original statement.
+
+For example:
+
+```cpp
+while (x == y)
+{
+  something();
+  somethingelse();
+}
+finalthing();
+```
+
+### File names
+
+File names should be lower case with underscore word separators. C file names use the `.c` extension, C++ files use the `.cc` extension and header files use the `.h` extension.
+
+### Language standards
+
+For pure-C files we use C99 and for C++ we use C++11. The code need to be able to compile on multiple platforms using different compilers (for example: Windows / Linux, x86_64 / ARM).
+
+### Line lengths
+
+Lines should be no more than 80 characters.
+
+### Comments
+
+Single line / inline code comments can use the double slash (`//`) style of coding, whereas multi-line code comments should use `/*` as a start and `*/` at the end, with the text indented by 2 spaces, for example:
+
+```cpp
+/*
+  This is a multi-line code comment.
+  It has an indentation of two spaces.
+*/
+```
+
+### Variables classes, and functions
+
+Variables and functions should be descriptive and in "snake case", for example:
+
+```cpp
+void my_function(uint16 variable_name)
+{
+```
+
+Class names should also be "snake case" but should start with an upper-case character. Such as this:
+
+```cpp
+class Buffered_logs
+{
+```
+
+### Constant integers
+
+Constant integers that are used to define elements such as buffer sizes should defined rather than used directly. This is because the integer could change and uses of it could be missed. For example:
+
+```cpp
+char *buffer= my_malloc(PSI_INSTRUMENT_ME, 1024, MYF(MY_WME));
+
+snprint(buffer, 1024, "%d: %s", integer, text);
+```
+
+Could become:
+
+```cpp
+const int buffer_size= 1024;
+char *buffer= my_malloc(PSI_INSTRUMENT_ME, buffer_size, MYF(MY_WME));
+
+snprint(buffer, buffer_size, "%d: %s", integer, text);
+```
+
+Alternatively the integer can be defined using an `enum` or `#define`.
+
+### Spacing
```

Comment:
It would be useful to recommend cleaning up whitespace in source files:

1. No trailing whitespace at end-of-line
2. No trailing blank lines at end-of-file
3. `\n` (POSIX) as end-of-line, not `\r\n` (Windows), for all text files
4. ***Some consistent standard for tabs vs. spaces for indentation*** so that all developers have a consistent view of the indentation of source files

I'm biased, but I'd also put in a plug for [dlenski/wtf](https://github.com/dlenski/wtf) as a tool that can do all these cleanups automatically, including [automatically doing them as part of a Git pre-commit hook](https://github.com/dlenski/wtf#git-pre-commit-hooks).

Reply:
I'll add these, but I don't want to add too much in this PR, otherwise it will never get merged. I want this to be more of a stub for other developers to add more things to.

Eventually I'd like to have a pre-commit hook and Buildbot test, it has not been decided which yet but I'll add that to the list. But I'd like narrower rules defined before we go that far, as well as some rules for some of the plugins. It will also likely require a clean-up pass of what we have. I'd also like to define as much as possible in a .editorconfig file too (mostly line endings and indentation) as many IDEs support this (and others can with a plugin), but plugins would need rule sets first.

This first step is intended to make things easier for new community members so that I have something to refer to. I literally wanted to refer contributors to this document a few times in the last few weeks to help them.